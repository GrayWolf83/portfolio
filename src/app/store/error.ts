import { AppDispatch, RootState } from './index'
import { createSlice, PayloadAction } from '@reduxjs/toolkit'

type errorState = {
	error: string | null
}

const initialState: errorState = {
	error: null,
}

const errorSlice = createSlice({
	name: 'errors',
	initialState,
	reducers: {
		loadingError(state, action: PayloadAction<string>) {
			state.error = action.payload
		},
		errorCleared(state) {
			state.error = null
		},
	},
})

const { loadingError, errorCleared } = errorSlice.actions

export const setLoadingError =
	(error: any) => async (dispatch: AppDispatch) => {
		if (error?.response?.data?.message) {
			dispatch(loadingError(error.response.data.message))
		}
	}

export const clearErrors = () => (dispatch: AppDispatch) => {
	dispatch(errorCleared())
}

export const getError = () => (state: RootState) => {
	return state.errors.error
}

export default errorSlice.reducer
