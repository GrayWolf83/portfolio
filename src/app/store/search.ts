import { AppDispatch, RootState } from './index'
import { setLoadingError } from './error'
import { createSlice, PayloadAction } from '@reduxjs/toolkit'
import { IUser } from '../types/IUser'
import userService from '../services/user.service'

type SearchState = {
	entities: IUser[]
	isLoading: boolean
	error: string | null
	dataLoaded: boolean
	page: number
	search: string
	usersCount: number
}

const initialState: SearchState = {
	entities: [],
	isLoading: false,
	error: null,
	dataLoaded: false,
	page: 1,
	search: '',
	usersCount: 0,
}

const searchSlice = createSlice({
	name: 'search',
	initialState,
	reducers: {
		serachLoadingStart: (state) => {
			state.isLoading = true
		},
		serachLoadingEnd: (state) => {
			state.isLoading = false
		},
		searchDataLoaded: (
			state,
			action: PayloadAction<{
				entities: IUser[]
				usersCount: number
				search: string
				page: number
			}>,
		) => {
			state.entities = action.payload.entities
			state.dataLoaded = true
			state.page = action.payload.page
			state.usersCount = action.payload.usersCount
			state.search = action.payload.search
		},
	},
})

const { serachLoadingStart, serachLoadingEnd, searchDataLoaded } =
	searchSlice.actions

export const searchUsersList =
	(search: string, page: number) => async (dispatch: AppDispatch) => {
		dispatch(serachLoadingStart())
		try {
			const payload = await userService.searchUsers(search, page)
			dispatch(
				searchDataLoaded({
					entities: payload.items,
					usersCount: payload.total_count,
					search,
					page,
				}),
			)
		} catch (error: any) {
			dispatch(setLoadingError(error))
		} finally {
			dispatch(serachLoadingEnd())
		}
	}

export const getUsersList = () => (state: RootState) => {
	return state.search.entities
}

export const getUsersCount = () => (state: RootState) => {
	return state.search.usersCount
}

export const getSearchLoadingStatus = () => (state: RootState) => {
	return state.search.isLoading
}

export const getSearchCurrentPage = () => (state: RootState) => {
	return state.search.page
}

export default searchSlice.reducer
