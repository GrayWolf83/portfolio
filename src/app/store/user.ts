import { IRepo } from '../types/IRepo'
import { AppDispatch, RootState } from './index'
import { setLoadingError } from './error'
import { createSlice, PayloadAction } from '@reduxjs/toolkit'
import { IUser } from '../types/IUser'
import userService from '../services/user.service'

type userState = {
	currentUser: IUser | null
	isLoading: boolean
	error: string | null
	dataLoaded: boolean
	repos: IRepo[]
	skills: string[]
}

const initialState: userState = {
	currentUser: null,
	isLoading: false,
	error: null,
	dataLoaded: false,
	repos: [],
	skills: [],
}

const userSlice = createSlice({
	name: 'user',
	initialState,
	reducers: {
		userLoadingStart: (state) => {
			state.isLoading = true
		},
		userLoadingEnd: (state) => {
			state.isLoading = false
		},
		userLoadingError: (state, action: PayloadAction<string>) => {
			state.error = action.payload
		},
		userLoaded: (state, action: PayloadAction<IUser>) => {
			state.currentUser = action.payload
			state.dataLoaded = true
		},
		userReposLoaded: (state, action: PayloadAction<IRepo[]>) => {
			state.repos = action.payload
		},
		userSkillsLoaded: (
			state,
			action: PayloadAction<{ [key: string]: number }[]>,
		) => {
			let skills = action.payload
				.map((item) => {
					return Object.keys(item).map((name) => name)
				})
				.flat()

			state.skills = Array.from(new Set(skills))
		},
	},
})

const {
	userLoadingStart,
	userLoadingEnd,
	userLoaded,
	userReposLoaded,
	userSkillsLoaded,
	userLoadingError,
} = userSlice.actions

export const loadUser = (login: string) => async (dispatch: AppDispatch) => {
	dispatch(userLoadingStart())
	try {
		const payload = await userService.getUser(login)
		dispatch(userLoaded(payload))

		const repos = await userService.getRepos(login)
		dispatch(userReposLoaded(repos))
	} catch (error: any) {
		dispatch(userLoadingError(error))
	} finally {
		dispatch(userLoadingEnd())
	}
}

export const loadUserSkills =
	(names: string[]) => async (dispatch: AppDispatch) => {
		dispatch(userLoadingStart())
		try {
			const skills = await userService.getSkills(names)
			dispatch(userSkillsLoaded(skills))
		} catch (error: any) {
			dispatch(userLoadingError(error))
		} finally {
			dispatch(userLoadingEnd())
		}
	}

export const getCurrentUser = () => (state: RootState) => {
	return state.user.currentUser
}

export const getUserRepos = () => (state: RootState) => {
	return state.user.repos
}

export const getUserSkills = () => (state: RootState) => {
	return state.user.skills
}

export const getUserLoadingStatus = () => (state: RootState) => {
	return state.user.isLoading
}

export default userSlice.reducer
