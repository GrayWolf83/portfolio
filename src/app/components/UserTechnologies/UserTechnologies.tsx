import { Box, Card, Grid, Typography } from '@mui/material'
import React from 'react'
import { useAppSelector } from '../../hooks/useAppReduxHooks'
import { getUserLoadingStatus, getUserSkills } from '../../store/user'
import Loader from '../Loader'

interface IUserTechnologies {}

const UserTechnologies: React.FC<IUserTechnologies> = () => {
	const skills = useAppSelector(getUserSkills())
	const isLoading = useAppSelector(getUserLoadingStatus())

	return (
		<Grid item md={8} xs={12} data-testid='UserTechnologies'>
			<Card
				sx={{
					height: '100%',
					width: '100%',
					p: '5px 0 5px',
				}}>
				<Typography
					variant='h6'
					color='primary'
					sx={{ paddingLeft: '5px' }}>
					Технологии:
				</Typography>
				{skills.length && !isLoading ? (
					<Grid container spacing={2}>
						{skills.map((skill) => (
							<Grid
								item
								sm={3}
								xs={6}
								key={skill}
								sx={{
									width: '100%',
									mt: '10px',
								}}>
								<Box
									sx={{
										width: '90px',
										height: '90px',
										border: '10px solid var(--primary-color)',
										borderRadius: '50%',
										display: 'flex',
										flexDirection: 'column',
										alignItems: 'center',
										justifyContent: 'center',
										margin: '0 auto',
									}}>
									<Typography
										variant='body1'
										color='primary'
										textAlign='center'>
										{skill}
									</Typography>
								</Box>
							</Grid>
						))}
					</Grid>
				) : (
					<Loader />
				)}
			</Card>
		</Grid>
	)
}

export default UserTechnologies
