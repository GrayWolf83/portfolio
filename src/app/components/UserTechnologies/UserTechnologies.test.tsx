import React from 'react'
import { render, screen } from '@testing-library/react'
import { BrowserRouter } from 'react-router-dom'
import { Provider } from 'react-redux'
import store from '../../store'
import UserTechnologies from './UserTechnologies'

describe('UserTechnologies component', () => {
	test('UserTechnologies render', () => {
		render(
			<Provider store={store}>
				<BrowserRouter>
					<UserTechnologies />
				</BrowserRouter>
			</Provider>,
		)

		const userTechnologies = screen.getByTestId('UserTechnologies')
		expect(userTechnologies).toBeInTheDocument()
	})

	test('UserTechnologies snapshot', () => {
		expect(
			render(
				<BrowserRouter>
					<UserTechnologies />
				</BrowserRouter>,
			),
		).toMatchSnapshot()
	})
})
