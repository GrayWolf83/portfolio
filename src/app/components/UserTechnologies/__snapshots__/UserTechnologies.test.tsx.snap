// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`UserTechnologies component UserTechnologies snapshot 1`] = `
Object {
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <div
        class="MuiGrid-root MuiGrid-item MuiGrid-grid-xs-12 MuiGrid-grid-md-4 css-13i4rnv-MuiGrid-root"
        data-testid="UserTechnologies"
      >
        <div
          class="MuiPaper-root MuiPaper-elevation MuiPaper-rounded MuiPaper-elevation1 MuiCard-root css-u5bj03-MuiPaper-root-MuiCard-root"
        >
          <h6
            class="MuiTypography-root MuiTypography-h6 css-khsj22-MuiTypography-root"
          >
            Технологии:
          </h6>
          <div
            class="MuiCardContent-root css-46bh2p-MuiCardContent-root"
          >
            <h6
              class="MuiTypography-root MuiTypography-h6 css-oblp1g-MuiTypography-root"
            >
              Технологии:
            </h6>
            <p
              class="MuiTypography-root MuiTypography-body1 css-vkc4v8-MuiTypography-root"
            >
              Нет данных для отображения
            </p>
          </div>
        </div>
      </div>
    </div>
  </body>,
  "container": <div>
    <div
      class="MuiGrid-root MuiGrid-item MuiGrid-grid-xs-12 MuiGrid-grid-md-4 css-13i4rnv-MuiGrid-root"
      data-testid="UserTechnologies"
    >
      <div
        class="MuiPaper-root MuiPaper-elevation MuiPaper-rounded MuiPaper-elevation1 MuiCard-root css-u5bj03-MuiPaper-root-MuiCard-root"
      >
        <h6
          class="MuiTypography-root MuiTypography-h6 css-khsj22-MuiTypography-root"
        >
          Технологии:
        </h6>
        <div
          class="MuiCardContent-root css-46bh2p-MuiCardContent-root"
        >
          <h6
            class="MuiTypography-root MuiTypography-h6 css-oblp1g-MuiTypography-root"
          >
            Технологии:
          </h6>
          <p
            class="MuiTypography-root MuiTypography-body1 css-vkc4v8-MuiTypography-root"
          >
            Нет данных для отображения
          </p>
        </div>
      </div>
    </div>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;
